File: src/android/app/src/main/java/info/cemu/Cemu/CemuApplication.java
package info.cemu.Cemu;

import android.app.Application;
import android.util.DisplayMetrics;

import java.io.File;
import java.io.PrintWriter;
import java.io.StringWriter;

import info.cemu.Cemu.nativeinterface.NativeEmulation;
import info.cemu.Cemu.nativeinterface.NativeGraphicPacks;
import info.cemu.Cemu.nativeinterface.NativeLogging;
import info.cemu.Cemu.nativeinterface.NativeSwkbd;

public class CemuApplication extends Application {
    static {
        System.loadLibrary("CemuAndroid");
    }

    private static Thread.UncaughtExceptionHandler defaultUncaughtExceptionHandler;
    private static CemuApplication application;

    public CemuApplication() {
        application = this;
    }

    public static CemuApplication getApplication() {
        return application;
    }

    public File getInternalFolder() {
        var externalFilesDir = getExternalFilesDir(null);
        if (externalFilesDir != null) {
            return externalFilesDir;
        }
        return getFilesDir();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        if (defaultUncaughtExceptionHandler == null) {
            defaultUncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler();
        }
        Thread.setDefaultUncaughtExceptionHandler((thread, exception) -> {
            StringWriter stringWriter = new StringWriter();
            PrintWriter printWriter = new PrintWriter(stringWriter);
            exception.printStackTrace(printWriter);
            String stacktrace = stringWriter.toString();
            NativeLogging.crashLog(stacktrace);
            defaultUncaughtExceptionHandler.uncaughtException(thread, exception);
        });
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        NativeEmulation.setDPI(displayMetrics.density);
        NativeEmulation.initializeActiveSettings(getInternalFolder().toString(), getInternalFolder().toString());
        NativeEmulation.initializeEmulation();
        NativeSwkbd.initializeSwkbd();
        NativeGraphicPacks.refreshGraphicPacks();
    }
}

File: src/android/app/src/main/java/info/cemu/cemu/CemuApplication.kt
package info.cemu.cemu

import android.app.Application
import info.cemu.cemu.nativeinterface.NativeEmulation.initializeActiveSettings
import info.cemu.cemu.nativeinterface.NativeEmulation.initializeEmulation
import info.cemu.cemu.nativeinterface.NativeEmulation.setDPI
import info.cemu.cemu.nativeinterface.NativeGraphicPacks.refreshGraphicPacks
import info.cemu.cemu.nativeinterface.NativeLogging.crashLog
import info.cemu.cemu.nativeinterface.NativeSwkbd.initializeSwkbd
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter

class CemuApplication : Application() {
    init {
        Application = this
    }

    val internalFolder: File
        get() {
            val externalFilesDir = getExternalFilesDir(null)
            if (externalFilesDir != null) {
                return externalFilesDir
            }
            return filesDir
        }

    override fun onCreate() {
        super.onCreate()
        if (DefaultUncaughtExceptionHandler == null) {
            DefaultUncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
        }
        Thread.setDefaultUncaughtExceptionHandler { thread: Thread, exception: Throwable ->
            val stringWriter = StringWriter()
            val printWriter = PrintWriter(stringWriter)
            exception.printStackTrace(printWriter)
            val stacktrace = stringWriter.toString()
            crashLog(stacktrace)
            DefaultUncaughtExceptionHandler!!.uncaughtException(
                thread,
                exception
            )
        }
        val displayMetrics = resources.displayMetrics
        setDPI(displayMetrics.density)
        initializeActiveSettings(internalFolder.toString(), internalFolder.toString())
        initializeEmulation()
        initializeSwkbd()
        refreshGraphicPacks()
    }

    companion object {
        init {
            System.loadLibrary("CemuAndroid")
        }

        private var DefaultUncaughtExceptionHandler: Thread.UncaughtExceptionHandler? = null

        @JvmStatic
        lateinit var Application: CemuApplication
            private set
    }
}
